shared-module/display/ili9341.c

```diff
@@ -0,0 +1,229 @@
+}
+
+inline void write_command(ili9341_obj_t *self, uint8_t c) {
+    common_hal_digitalio_digitalinout_set_value(self->dc, false);
+
+    spi_write(self, &c, sizeof(c));
+}
+
+inline void write_data(ili9341_obj_t *self, uint8_t c) {
+    common_hal_digitalio_digitalinout_set_value(self->dc, true);
+
+    spi_write(self, &c, sizeof(c));
+}
+
+void display_init9341_write_init(ili9341_obj_t *self) {
+    write_command(self, ILI9341_SWRESET);
+    mp_hal_delay_ms(120);
+
+    write_command(self, ILI9341_DISPOFF);
+    mp_hal_delay_ms(120);
+
+    write_command(self, ILI9341_PWCTRB);
+    write_data(self, 0x00);
+    write_data(self, 0XC1);
+    write_data(self, 0X30);
+
+    write_command(self, ILI9341_TIMCTRA);
+    write_data(self, 0x85);
+    write_data(self, 0x00);
+    write_data(self, 0x78);
+
+    write_command(self, ILI9341_PWCTRSEQ);
+    write_data(self, 0x39);
+    write_data(self, 0x2C);
+    write_data(self, 0x00);
+    write_data(self, 0x34);
+    write_data(self, 0x02);
+
+    write_command(self, ILI9341_PUMP);
+    write_data(self, 0x20);
+
+    write_command(self, ILI9341_TIMCTRB);
+    write_data(self, 0x00);
+    write_data(self, 0x00);
+
+    write_command(self, ILI9341_PWCTR1);
+    write_data(self, 0x23);
+
+    write_command(self, ILI9341_PWCTR2);
+    write_data(self, 0x10);
+
+    write_command(self, ILI9341_VMCTR1);
+    write_data(self, 0x3e);
+    write_data(self, 0x28);
+
+    write_command(self, ILI9341_VMCTR2);
+    write_data(self, 0x86);
+
+    write_command(self, ILI9341_MADCTL);
+    write_data(self, 0x48);
+
+    write_command(self, ILI9341_PIXFMT);
+    write_data(self, 0x55);
+
+    write_command(self, ILI9341_FRMCTR1);
+    write_data(self, 0x00);
+    write_data(self, 0x18);
+
+    write_command(self, ILI9341_DFUNCTR);
+    write_data(self, 0x08);
+    write_data(self, 0x82);
+    write_data(self, 0x27);
+
+    write_command(self, ILI9341_ENGMCTR);
+    write_data(self, 0x00);
+
+    write_command(self, ILI9341_GAMMASET);
+    write_data(self, 0x01);
+
+    write_command(self, ILI9341_GMCTRP1);
+    write_data(self, 0x0F);
+    write_data(self, 0x31);
+    write_data(self, 0x2B);
+    write_data(self, 0x0C);
+    write_data(self, 0x0E);
+    write_data(self, 0x08);
+    write_data(self, 0x4E);
+    write_data(self, 0xF1);
+    write_data(self, 0x37);
+    write_data(self, 0x07);
+    write_data(self, 0x10);
+    write_data(self, 0x03);
+    write_data(self, 0x0E);
+    write_data(self, 0x09);
+    write_data(self, 0x00);
+
+    write_command(self, ILI9341_GMCTRN1);
+    write_data(self, 0x00);
+    write_data(self, 0x0E);
+    write_data(self, 0x14);
+    write_data(self, 0x03);
+    write_data(self, 0x11);
+    write_data(self, 0x07);
+    write_data(self, 0x31);
+    write_data(self, 0xC1);
+    write_data(self, 0x48);
+    write_data(self, 0x08);
+    write_data(self, 0x0F);
+    write_data(self, 0x0C);
+    write_data(self, 0x31);
+    write_data(self, 0x36);
+    write_data(self, 0x0F);
```

Comment:
All this initialization could be stored in a string/array, and executed in a loop, instead of individual commands — the resulting code will be much smaller then.

Reply:
Sure, will do.

Reply:
Why bother putting this init here? It doesn't need to be fast so it can be in Python.

Instead, removing it can make this class more generic. Suppose it just manages the buffer and the column and row setting. Then it can work with more SPI displays that use the same approach but have slightly different command values.

Reply:
Not necessarily. A board can choose to freeze python libs in or ship with them on the drive.